core/src/main/java/org/elasticsearch/common/xcontent/support/AbstractXContentParser.java

```diff
@@ -289,7 +289,7 @@ public BytesRef utf8BytesOrNull() throws IOException {
             token = parser.nextToken();
         }
         ArrayList<Object> list = new ArrayList<>();
-        for (; token != XContentParser.Token.END_ARRAY; token = parser.nextToken()) {
+        for (; token != null && token.isValue(); token = parser.nextToken()) {
             list.add(readValue(parser, mapFactory, token));
```

Comment:
Yes, the condition `token != null && token.isValue()` is not enough, we need to consume all the parser and stop when we hit the end of the array or the end of the parent's object.